Subscriber/Cronjob/ProcessNotifications.php

```diff
@@ -59,9 +80,13 @@ public static function getSubscribedEvents()
      */
     public function runCronjob(Shopware_Components_Cron_CronJob $job)
     {
+        $textNotifications = $this->fifoTextNotificationLoader->get();
+
+        $this->incomingNotificationManager->convertNotifications($textNotifications);
```

Comment:
@acampos1916 

* The convertNotifications is a Generator and it needs to be consumed in a foreach or with iterator_to_array for the method to be executed.  
  
* I was thinking of a possible feature improvement  to add a `status` to the `TextNotification` entity
Example values: RECEIVED | ERROR | CONVERTED

  And have an automated clean-up (status CONVERTED or PROCESSED and updated >= 6 months ago or a reasonable timing so that this table does not just grow over the years.
  (The same could be done for the converted notifications as well)

Reply:
Indeed the status and cleanup should be done for `TextNotification`. I'll create a ticket for this.

In this PR the `convertNotifications` does not return a generator, maybe you're confused? What's your opinion on its current implementation BTW?

Reply:
Right `convertNotifications` was re-written, never mind my comment about the generator.

I dug a bit into the code base and do have some concerns about  
the handling of Incoming notifications in the long run:  

* the current in place handling needs to be improved,
* the added table by PR adds complexity that might be avoided.

Ideally the incoming notification should be put on a job queue (running in a separate process) after the authorization is valid,  so `'[accepted]'` is sent as fast as possible.

For the saving, converting, processing the Incoming Notifications,  
there is a lot done before knowing if the notification is relevant or not:

1. saving the incoming notification
1. convert via cron
1. process the converted notification

If for example only one notification endpoint with one Shop is used both the DB table of text notification and converted notifications would be near identical.  

It also adds complexity before handling a notification: now 2 crons run for converting and handling the notification.  
This could be managed by one table to ease on the complexity during runtime (only one table to verify and one cron that processes them).

As for the current Generator that only yields on exceptions would better always yield preferable an object.
The generator could move to (or a better spot) `Shopware_Controllers_Frontend_Notification::getNotificationItems`
and in that method create the Notification Instance (`\AdyenPayment\Components\Builder\NotificationBuilder`) without validation, the method would then yield `\AdyenPayment\Models\Notification` objects to be used further along the process.

Secondly we would need some improved handling if we can determine before saving, 
if the notification should be saved in the first place.

An implementation a bit based on the Authorization validator:

1. a main service: SkipIncomingNotificationRunner (or some fitting name)  
with service arguments (with highest executed first)

    * SkipNotRelevantForSystem (example: notification meant for an ERP)
    * SkipReportNotification
    * Skip ...  

    The "Skip" services should then relative fast determine if the message should be skipped and return true if it should. This would halt the execution of the runner and continue if there is another notification in the request.

    * When passing (notification is not skipped) the runner then only saves the notification to DB.

    * When the notification should be skipped a decorator SkipIncomingNotificationRunnerDecorator) could then log the notification
      * obfuscated by a formatter or so to hide sensitive data
      * this way for debugging the notification is stored in the log files  
    and with log rotation they will be cleaned automatically

1. A further improvement could be taken on Processing.  
A similar approach if the saved notification can be processed or not and updating a status flag on the saved notification.
Example SkipNotificationWithoutOrder (this skip would require a DB lookup and is betted done during processing an not during receiving as it would take longer)

1. a clean-up cron could be made to remove unprocessable notifications (after a decent time span)

In this is a rough way we could ease the load on additional data in the DB and 
have a more robust / extendable process of Notifications in the long run.

Reply:
Thanks for the feedback @bortefi! Indeed in all plugins we store the incoming notification directly in the notifications table. Since the Notifications entity in this plugin links to the order we decided to create a separate entity for now. But this write-up is very valuable and we'll keep it in mind.